homeassistant/components/network/network.py

```diff
@@ -0,0 +1,237 @@
+        await self.async_load()
+        self._adapters = load_adapters(self._next_broadcast_hop)
+
+    async def async_migrate_from_zeroconf(self, zc_config) -> None:
+        """Migrate configuration from zeroconf."""
+        if self._data:
+            return
+
+        if (
+            ZC_CONF_DEFAULT_INTERFACE in zc_config
+            and not zc_config[ZC_CONF_DEFAULT_INTERFACE]
+        ):
+            self._data[ATTR_CONFIGURED_ADAPTERS] = _adapters_with_exernal_addresses(
+                self._adapters
+            )
+            await self._async_save()
+
+    @callback
+    def async_configure(self) -> None:
+        """Configure from storage."""
+        if not _enable_adapters(self._adapters, self.configured_adapters):
+            _auto_detect_adapters(self._adapters)
+
+    async def async_reconfig(self, config: dict[str, Any]) -> None:
+        """Reconfigure network."""
+        config = NETWORK_CONFIG_SCHEMA(config)
+        self._data[ATTR_CONFIGURED_ADAPTERS] = config[ATTR_CONFIGURED_ADAPTERS]
+        for adapter in self._adapters:
+            adapter["enabled"] = False
+        _enable_adapters(self._adapters, self.configured_adapters)
+        await self._async_save()
+
+    async def async_load(self) -> None:
+        """Load config."""
+        if stored := await self._store.async_load():
+            self._data = cast(dict, stored)
+
+    async def _async_save(self) -> None:
+        """Save preferences."""
+        await self._store.async_save(self._data)
+
+
+def _enable_adapters(adapters: list[Adapter], enabled_interfaces: list[str]) -> bool:
+    """Enable configured adapters."""
+    if not enabled_interfaces:
+        return False
+
+    found_adapter = False
+    for adapter in adapters:
+        if adapter["name"] in enabled_interfaces:
+            adapter["enabled"] = True
+            found_adapter = True
+
+    return found_adapter
+
+
+def _auto_detect_adapters(adapters: list[Adapter]) -> None:
+    """Enable the default adapter or all adapters with external addresses."""
+    for adapter in adapters:
+        if adapter["default"] and _adapter_has_external_address(adapter):
+            adapter["enabled"] = True
+            return
+
+    _enable_adapters(adapters, _adapters_with_exernal_addresses(adapters))
+
+
+def _adapters_with_exernal_addresses(adapters: list[Adapter]) -> list[str]:
+    """Enable all interfaces with an external address."""
+    return [
+        adapter["name"]
+        for adapter in adapters
+        if _adapter_has_external_address(adapter)
+    ]
+
+
+def _adapter_has_external_address(adapter: Adapter) -> bool:
+    """Adapter has a non-loopback and non-link-local address."""
+    return any(
+        _has_external_address(v4_config["address"]) for v4_config in adapter["ipv4"]
+    ) or any(
+        _has_external_address(v6_config["address"]) for v6_config in adapter["ipv6"]
+    )
+
+
+def _has_external_address(ip_str: str):
```

Comment:
Maybe add it to `.strict-typing` as you have it mostly annotated already?
```suggestion
def _has_external_address(ip_str: str) -> bool:
```